pandas/_libs/tslibs/conversion.pyx

```diff
@@ -608,9 +608,66 @@ cpdef inline datetime localize_pydatetime(datetime dt, object tz):
 # ----------------------------------------------------------------------
 # Timezone Conversion
 
+cdef inline int64_t[:] _tz_convert_dst(ndarray[int64_t] values, tzinfo tz,
+                                       bint to_utc=True):
+    """
+    tz_convert for non-UTC non-tzlocal cases where we have to check
+    DST transitions pointwise.
+
+    Parameters
+    ----------
+    values : ndarray[int64_t]
+    tz : tzinfo
+    to_utc : bool
+        True if converting _to_ UTC, False if converting _from_ utc
+
+    Returns
+    -------
+    result : ndarray[int64_t]
+    """
+    cdef:
+        Py_ssize_t n = len(values)
+        Py_ssize_t i, j, pos
+        ndarray[int64_t] result = np.empty(n, dtype=np.int64)
+        ndarray[int64_t] tt, trans, deltas
+        ndarray[Py_ssize_t] posn
+        int64_t v
+
+    trans, deltas, typ = get_dst_info(tz)
+    if not to_utc:
+        # We add `offset` below instead of subtracting it
+        deltas = -1 * deltas
+
+    tt = values[values != NPY_NAT]
+    if not len(tt):
+        # if all NaT, return all NaT
+        return values
+
+    posn = trans.searchsorted(tt, side='right')
```

Comment:
The other places where `get_dst_info` is used don't do this `tt` thing.

`_normalize_local`, `period.localize_dt64_arr_to_period`, `resolution._reso_local` all do:

```
trans, deltas, typ = get_dst_info(tz)
_pos = trans.searchsorted(stamps, side='right') - 1
 if _pos.dtype != np.int64:
        _pos = _pos.astype(np.int64)
pos = _pos
```

`is_date_array_normalized` and `tslib.ints_to_pydatetime` makes `searchsorted` calls pointwise

```
trans, deltas, typ = get_dst_info(tz)
for i in range(n):
    pos = trans.searchsorted(stamps[i]) - 1
```

I'm checking to see if anything breaks if we change it here to match the others.

Perf Considerations:
- The `tt` thing here presumably improves perf in cases with many `NaT`s
- Without the `values != NPY_NAT` check we could pass `values` as an `int64_t[:]` which I _think_ would be more performant.
- There is a `cnp.PY_ArraySearchSorted` that could be used to avoid reaching into python-space, but apparently there is a bug in cython's numpy.pxd that makes this non-trivial (https://stackoverflow.com/questions/28184211/calling-pyarray-searchsorted-from-cython-3-or-4-arguments?rq=1)

Reply:
> Without the values != NPY_NAT check we could pass values as an int64_t[:] which I think would be more performant.

why? NPY_NAT is i8

Reply:
not sure why you can search an array of timestamps to searchsorted here?

Reply:
> why? NPY_NAT is i8

AFAICT `int64_t[:]` doesn't support `__eq__`.  You can basically iterate through it and little else.

Reply:
> not sure why you can search an array of timestamps to searchsorted here?

I don't understand the question.  `trans.searchsorted(x)` assumes (doesn't check) that `trans` is sorted, doesn't care about `x` (which can be array or scalar)